File: backend/src/api/library.rs
use axum::{Json, extract::{State, Path, Query}};
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use crate::{app::AppState, db::queries, error::AppResult};

#[derive(Debug, Deserialize)]
pub struct LibraryQuery {
    pub page: Option<i64>,
    pub limit: Option<i64>,
    pub r#type: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct RequeueResponse {
    pub job_id: String,
    pub status: String,
}

pub async fn list_library(
    State(state): State<Arc<AppState>>,
    Query(params): Query<LibraryQuery>,
) -> AppResult<Json<queries::LibraryResponse>> {
    let page = params.page.unwrap_or(1).max(1);
    let limit = params.limit.unwrap_or(20).min(100);
    let media_type = params.r#type.as_deref();

    let response = queries::get_completed_jobs_grouped(&state.db, media_type, page, limit).await?;
    Ok(Json(response))
}

pub async fn requeue_job(
    State(state): State<Arc<AppState>>,
    Path(job_id): Path<String>,
) -> AppResult<Json<RequeueResponse>> {
    queries::requeue_job(&state.db, &job_id).await?;

    Ok(Json(RequeueResponse {
        job_id,
        status: "queued".to_string(),
    }))
}

File: backend-bun/src/api/library.ts
import { toSnake } from "../db/transform";
import type { Context } from "hono";
import type { AppBindings } from "../app";
import { notFound } from "../error";
import * as queries from "../db/queries";

export async function listLibrary(c: Context<AppBindings>) {
  const { db } = c.var;
  const page = Math.max(1, parseInt(c.req.query("page") || "1", 10));
  const limit = Math.min(100, parseInt(c.req.query("limit") || "20", 10));
  const mediaType = c.req.query("type") || null;

  const result = queries.getCompletedJobsGrouped(db, mediaType, page, limit);
  return c.json(toSnake(result));
}

export async function requeueJobHandler(c: Context<AppBindings>) {
  const { db } = c.var;
  const jobId = c.req.param("jobId")!;

  const ok = queries.requeueJob(db, jobId);
  if (!ok) throw notFound("Job not found");

  return c.json({ job_id: jobId, status: "queued" });
}

export async function getLibraryItem(c: Context<AppBindings>) {
  const { db } = c.var;
  const imdbId = c.req.param("imdbId")!;

  const detail = queries.getLibraryDetail(db, imdbId);
  return c.json(toSnake(detail));
}

export async function deleteLibraryItem(c: Context<AppBindings>) {
  const { db } = c.var;
  const imdbId = c.req.param("imdbId")!;

  const count = queries.deleteJobsByImdbId(db, imdbId);
  if (count === 0) throw notFound("No completed jobs for " + imdbId);

  return c.json({ deleted: count });
}
